client/rpc/src/eth_pubsub.rs

```diff
@@ -384,116 +347,92 @@ where
-									.map(|res| {
-										UniqueSaturatedInto::<u64>::unique_saturated_into(res)
-									});
-								// Best imported block.
-								let current_block =
-									UniqueSaturatedInto::<u64>::unique_saturated_into(
-										client.info().best_number,
-									);
+					// Gets the node syncing status.
+					// The response is expected to be serialized either as a plain boolean
+					// if the node is not syncing, or a structure containing syncing metadata
+					// in case it is.
+					async fn status<C: HeaderBackend<B>, B: BlockT, H: ExHashT + Send + Sync>(
+						client: Arc<C>,
+						network: Arc<NetworkService<B, H>>,
+						starting_block: u64,
+					) -> PubSubSyncStatus {
+						if network.is_major_syncing() {
+							// Get the target block to sync.
+							// This value is only exposed through substrate async Api
+							// in the `NetworkService`.
+							let highest_block = network
+								.status()
+								.await
+								.ok()
+								.and_then(|res| res.best_seen_block)
+								.map(UniqueSaturatedInto::<u64>::unique_saturated_into);
+							// Best imported block.
+							let current_block = UniqueSaturatedInto::<u64>::unique_saturated_into(
+								client.info().best_number,
+							);
 
-								PubSubSyncStatus::Detailed(SyncStatusMetadata {
-									syncing: true,
-									starting_block,
-									current_block,
-									highest_block,
-								})
-							} else {
-								PubSubSyncStatus::Simple(false)
-							}
+							PubSubSyncStatus::Detailed(SyncStatusMetadata {
+								syncing: true,
+								starting_block,
+								current_block,
+								highest_block,
+							})
+						} else {
+							PubSubSyncStatus::Simple(false)
 						}
-						// On connection subscriber expects a value.
-						// Because import notifications are only emitted when the node is synced or
-						// in case of reorg, the first event is emited right away.
-						let _ = sink
-							.feed(Ok(PubSubResult::SyncState(
-								status(Arc::clone(&client), Arc::clone(&network), starting_block)
-									.await,
-							)))
-							.await;
+					}
+					// On connection subscriber expects a value.
+					// Because import notifications are only emitted when the node is synced or
+					// in case of reorg, the first event is emited right away.
+					let _ = sink.send(&PubSubResult::SyncState(
+						status(Arc::clone(&client), Arc::clone(&network), starting_block).await,
+					));
 
-						// When the node is not under a major syncing (i.e. from genesis), react
-						// normally to import notifications.
-						//
-						// Only send new notifications down the pipe when the syncing status changed.
-						client
+					// When the node is not under a major syncing (i.e. from genesis), react
+					// normally to import notifications.
+					//
+					// Only send new notifications down the pipe when the syncing status changed.
+					use std::sync::atomic::{AtomicBool, Ordering};
+					let last_syncing_status =
```

Comment:
Atomics are heavy-weaponry, how about doing something like this instead: https://github.com/humanode-network/frontier/commit/e93977a8cc9c0e2ebd7cf9e2c637a00393748d33

Reply:
Yes, it would work, thanks for the suggestion :)